### PR TITLE
Reimplement aircraft carriers and rocket launchers from Red Alert 2

### PR DESCRIPTION
**Pull request by ZivDero**
This PR reimplements SpawnManagerClass from Red Alert 2, used for aircraft carriers and rocket launchers.

In `RULES.INI`:
```ini
[SOMETECHNO]        ; TechnoType
Spawns=             ; AircraftType, the type that this Techno spawns.
SpawnsNumber=0      ; integer, the number of aircraft that this Techno spawns.
SpawnRegenRate=0    ; integer, the time it takes for a spawn to regenerate after death.
SpawnReloadRate=0   ; integer, the time it takes for a spawn to reload its ammo and restore its strength.
```

- Additionally, it's possible to specify an alternative voxel model to use when the spawner has no spawns docked.

In `RULES.INI`:
```ini
[SOMETECHNO]        ; TechnoType
NoSpawnAlt=no       ; boolean, should this Techno use an alternate voxel model when it's out of spawns?
                    ; When true, the model named SOMETECHNOWO will be used when it's out of spawns.
```

```{note}
`NoSpawnAlt` is only available for Technos that use voxel models.
```

```{note}
Unlike in Red Alert 2, the voxel model used by `NoSpawnAlt` is loaded into a separate area of memory from the turet model. This means that Technos that have `NoSpawnAlt=yes` set **can** have turrets.
```

- For the unit to create spawns upon attack, a flag needs to be set on its weapon.

In `RULES.INI`:
```ini
[SOMEWEAPON]  ; WeaponType
Spawner=no    ; boolean, does this weapon spawn aircraft when firing?
```

#### Rockets

- For rocket launchers to function, RocketTypes need to be defined.

In `RULES.INI`:
```ini
[RocketTypes]
0=SOMEROCKET

[SOMEROCKET]            ; RocketType
PauseFrames=0           ; integer, how many frames the rocket pauses on the launcher before tilting.
TiltFrames=60           ; integer, how many frames it takes for the rocket to tilt to firing position.
PitchInitial=0.21       ; float, starting pitch of the rocket before tilting up (0 = horizontal, 1 = vertical).
PitchFinal=0.5          ; float, ending pitch of the rocket after tilting up before it fires.
TurnRate=0.05           ; float, pitch maneuverability of rocket in air.
RaiseRate=1             ; integer, how much the missile will raise each turn on the launcher (for Cruise Missiles only).
Acceleration=0.4        ; float, this much is added to the rocket's velocity each frame during launch.
Altitude=768            ; integer, cruising altitude in leptons (1/256 of a cell): at this height rocket begins leveling off.
Damage=200              ; integer, the rocker does this much damage when it explodes.
EliteDamage=400         ; integer, the rocker does this much damage when it explodes when the spawner is elite.
BodyLength=256          ; integer, the length of  thebody of the rocket in leptons (1/256 of a cell).
LazyCurve=yes           ; boolean, is the rocket's path a big, lazy curve, like the V3 is Red Alert 2.
CruiseMissile=no        ; boolean, is this rocket a Cruise Missile, like Boomer missiles in Yuri's Revenge.
Type=                   ; AircraftType, the type this rocket is associated with.
Warhead=                ; WarheadType, the warhead that this rocket's explosion uses.
EliteWarhead=           ; WarheadType, the warhead that this rocket's explosion uses when the spawner is elite.
TakeoffAnim=            ; AnimType, the takeoff animation used by this rocket.
TrailAnim=              ; AnimType, the trail animation used by this rocket.
```

```{note}
The rocket object is an AircraftType, like in Red Alert 2. When determining its characteristics, the rocket will use the first RocketType whose `Type=` is equal to the type of the rocket itself.
```

- Additionally, for rocket launchers that launch two rockets at a time, a second spawn offset can de defined.

In `ART.INI`:
```ini
[SOMETECHNO]             ; TechnoType
SecondSpawnOffset=0,0,0  ; 3 integers, an offset to be added to the firing FLH for the second missile's location.
```

### CODE REVIEW

File: src/new/rockettype/rockettype.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+/*                 O P E N  S O U R C E  --  V I N I F E R A                  **
+/*******************************************************************************
+ *
+ *  @project       Vinifera
+ *
+ *  @file          ROCKETTYPE.CPP
+ *
+ *  @authors       ZivDero
+ *
+ *  @brief         Class containing condiguration for AircraftType rockets.
```
Comment (Rampastring) [inline src/new/rockettype/rockettype.cpp:11]:
```suggestion
 *  @brief         Class containing configuration for AircraftType rockets.
```

File: src/new/spawnmanager/spawnmanager_hooks.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        JMP(0x0047FE39);
+    }
+
+    extension = Extension::Fetch<TechnoClassExtension>(linked_to);
+    if (extension->SpawnManager && extension->SpawnManager->Preparing_Count())
+    {
+        _asm popad
+        // return 1;
+        JMP(0x0047FE39);
+    }
```
Comment (Rampastring) [inline src/new/spawnmanager/spawnmanager_hooks.cpp:95]: I can't imagine the performance improvement from this being significant enough to warrant the code duplication, especially as it'd be in the relatively uncommon "EMP'd unit" case. And I wouldn't be surprised if the compiler ended up generating the same code anyway - avoiding fetching the extension unless necessary.

```suggestion
    extension = Extension::Fetch<TechnoClassExtension>(linked_to);
    if (linked_to->EMPFramesRemaining > 0 || (extension->SpawnManager && extension->SpawnManager->Preparing_Count()))
    {
        _asm popad
        // return 1;
        JMP(0x0047FE39);
    }
```

File: src/new/spawnmanager/spawnmanager_hooks.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+
+/**
+ *  Patch to perform kamikaze AI.
+ *
+ *  @author: ZivDero
+ */
+DECLARE_PATCH(_LogicClass_AI_Kamikaze_AI_Patch)
+{
+    // Stolen instrution
```
Comment (Rampastring) [inline src/new/spawnmanager/spawnmanager_hooks.cpp:110]:
```suggestion
    // Stolen instruction
```

File: src/new/spawnmanager/spawnmanager.cpp [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+        case SpawnControlStatus::Preparing:
+            {
+                /**
+                 *  Missiles don't do this.
+                 */
+                if (control->IsSpawnedMissile)
+                    break;
+
+                /**
+                 *  If there's not target, return to base.
```
Comment (Rampastring) [inline src/new/spawnmanager/spawnmanager.cpp:446]:
```suggestion
                 *  If there's no target, return to base.
```

This comment seems a bit weird btw. The comparison below is `if (Target != nullptr)`. It seems more like "if there's a target, return to base"?

**Review by Rampastring — CHANGES_REQUESTED** [review]

File: src/new/locomotion/rocketlocomotion.cpp [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+ */
+IFACEMETHODIMP_(Matrix3D) RocketLocomotionClass::Draw_Matrix(int *key)
+{
+    Matrix3D matrix;
+    matrix.Make_Identity();
+
+    /**
+     *  Rotate the rocket to its current facing.
+     *  YR subtracts 8 from the facing, equivalent to 90 degrees, but for some reason
+     *  it's necessary to subtract 12 (135 degrees) in YR to achieve the same result.
```
[the remark this answers is not in the document]
Comment (ZivDero) [reply src/new/locomotion/rocketlocomotion.cpp:136]: Outdated comment from back when DirStruct stuff was broken, will fix.

File: src/extensions/aircraft/aircraftext_hooks.cpp [hunk 6]
```diff
@@ -37,14 +38,253 @@
...
+                }
+            }
+
+            if (Get_Weapon()->Weapon) {
+
+                /**
+                 *  Weapon equipped helicopters that run out of ammo and were
+                 *  brought in as reinforcements will leave the map.
+                 */
+                if (IsALoaner) {
```
[the remark this answers is not in the document]
Comment (ZivDero) [reply src/extensions/aircraft/aircraftext_hooks.cpp:198]: The code matches as far as I can see, tho...

File: src/extensions/building/buildingext_hooks.cpp [hunk 7]
```diff
@@ -934,6 +935,26 @@ DECLARE_PATCH(_EventClass_Execute_Archive_Selling_Patch)
...
+ *
+ *  @author: ZivDero
+ */
+DECLARE_PATCH(_BuildingClass_Captured_IncrementUnitTracker_Patch)
+{
+    GET_REGISTER_STATIC(BuildingClass*, this_ptr, esi);
+    static BuildingTypeClassExtension* ext;
+
+    ext = Extension::Fetch<BuildingTypeClassExtension>(this_ptr->Class);
+    if ((Session.Type == GAME_INTERNET || Session.Type == GAME_IPX) && !ext->IsDontScore)
```
[the remark this answers is not in the document]
Comment (ZivDero) [reply src/extensions/building/buildingext_hooks.cpp:949]: You misread the order of operations. Scoring only happens in Internet and Skirmish (vanilla is Internet only, changed for ts-patches), and only for not DontScore technos.

[the remark this answers is not in the document]
Comment (ZivDero) [reply src/extensions/building/buildingext_hooks.cpp:949]: Now, the addresses might contains a mistake tho...

**Review by Rampastring — APPROVED** [review]

**Pull request merged**